src/components/Tokens/TokenTable/TokenTable.tsx

```diff
@@ -149,14 +167,94 @@ export function LoadingTokenTable() {
   )
 }
 
-export default function TokenTable({ data }: { data: TokenData[] | undefined }) {
+interface TokenRowProps {
+  data: TokenData[]
+  index: number
+  style: CSSProperties
+}
+
+export default function TokenTable() {
   const showFavorites = useAtomValue<boolean>(showFavoritesAtom)
   const timePeriod = useAtomValue<TimePeriod>(filterTimeAtom)
-  const filteredTokens = useFilteredTokens(data)
-  const sortedFilteredTokens = useSortedTokens(filteredTokens)
+
+  const [page, setPage] = useState<number>(1)
+  const [topTokens, setTopTokens] = useState<TokenData[]>(
+    Array.from({ length: MAX_TOKENS_TO_LOAD }).map((_) => ({ loading: true } as TokenData))
+  )
+  const [error, setError] = useState<any>(undefined)
+  // const filteredTokens = useFilteredTokens(data)
+  // const sortedFilteredTokens = useSortedTokens(filteredTokens)
+
+  const isItemLoaded = (index: number) => index < topTokens.length && topTokens[index].loading === false
+
+  const loadMoreItems = (startIndex: number, stopIndex: number) => {
+    if (stopIndex >= page * PAGE_SIZE && page < MAX_PAGE) setPage(page + 1)
+    return fetchQuery<TopTokenQueryType>(environment, query, {
+      pageSize: PAGE_SIZE,
+      page,
+    })
+      .toPromise()
+      .then((data) => {
```

Comment:
i generally prefer async/await vs .then - is this because you have to specifically return a promise?

Reply:
yeah we need to return a promise to the infinite loader